Src/Base/AMReX_GpuContainers.H

```diff
@@ -52,16 +52,169 @@ namespace Gpu {
     using PolymorphicDeviceVector = PODVector<T>;
 #endif
 
+    static struct HostToDevice {} hostToDevice;
+    static struct DeviceToHost {} deviceToHost;
+    static struct DeviceToDevice {} deviceToDevice;
+
+    /**
+     * \brief A host-to-device copy routine. Note this is just a wrapper around
+     * memcpy, so it assumes contiguous storage. The amrex-provided containers
+     * like Gpu::HostVector, Gpu::DeviceVector, etc. meet this requirement.
+     *
+     * This version is blocking - CPU execution will halt until the copy is finished.
+     *
+     * \tparam InIter  The number of extra Real components in the particle struct
+     * \tparam OutIter The number of extra integer components in the particle struct
+     * \param HostToDevice A tag indicating that the copy is from the host to the device
+     * \param begin Where in the input to start reading
+     * \param end Where in the input to stop reading
+     * \param result Where in the output to start writing
+     *
+     * Example usage:
+     *     
+     *    Gpu::copy(Gpu::hostToDevice, a.begin(), a.end(), b.begin());
+     */
     template<class InIter, class OutIter>
-    OutIter thrust_copy (InIter begin, InIter end, OutIter result)
+    void copy (HostToDevice, InIter begin, InIter end, OutIter result) noexcept
     {
-#ifdef AMREX_USE_GPU
-        return thrust::copy(begin, end, result);
-#else
-        return std::copy(begin, end, result);
-#endif
+        using value_type = typename std::iterator_traits<InIter>::value_type;
+        auto size = std::distance(begin, end);
+        htod_memcpy(&(*result), &(*begin), size*sizeof(value_type));
```

Comment:
Maybe we can add a static_assert making sure `vaule_type` is trivially copyable (or is thrust complex)?